File: services/client/src/protocol/protocol.go
package protocol

import (
	"encoding/binary"
	"io"

	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/safe_socket"
)

const (
	MsgBet     byte = 1
	MsgAck     byte = 2
	MsgEndBets byte = 3
	MsgWinners byte = 4
)

const HeaderSize = 5

func SendMsg(socket io.Writer, msgType byte, payload []byte) error {
	header := make([]byte, HeaderSize)
	header[0] = msgType
	binary.BigEndian.PutUint32(header[1:], uint32(len(payload)))

	if err := safe_socket.SendAll(socket, header); err != nil {
		return err
	}
	if len(payload) > 0 {
		return safe_socket.SendAll(socket, payload)
	}
	return nil
}

func RecvMsg(socket io.Reader) (byte, []byte, error) {
	header, err := safe_socket.RecvAll(socket, HeaderSize)
	if err != nil {
		return 0, nil, err
	}
	msgType := header[0]
	length := binary.BigEndian.Uint32(header[1:])

	if length == 0 {
		return msgType, []byte{}, nil
	}

	payload, err := safe_socket.RecvAll(socket, int(length))
	if err != nil {
		return 0, nil, err
	}
	return msgType, payload, nil
}

File: services/server/src/protocol/protocol.py
import struct
import socket
from safe_socket import send_all, recv_all

MSG_BET = 1
MSG_ACK = 2
MSG_END_BETS = 3
MSG_WINNERS = 4

# Header: 1 byte msg_type (uint8) + 4 bytes length (uint32 big-endian) = 5 bytes
_HEADER_FORMAT = "!BI"
_HEADER_SIZE = struct.calcsize(_HEADER_FORMAT)


def send_msg(sock: socket.socket, msg_type: int, payload: bytes = b"") -> None:
    """Empaqueta la cabecera y envía el mensaje completo por el socket."""
    header = struct.pack(_HEADER_FORMAT, msg_type, len(payload))
    send_all(sock, header + payload)


def recv_msg(sock: socket.socket) -> tuple[int, bytes] | tuple[None, None]:
    """Recibe un mensaje con cabecera de tipo y longitud fija.

    Retorna (msg_type, payload). Si la conexión se cerró limpiamente al inicio,
    retorna (None, None).
    """
    header_bytes = recv_all(sock, _HEADER_SIZE)
    if not header_bytes:
        return None, None
    msg_type, length = struct.unpack(_HEADER_FORMAT, header_bytes)
    payload = recv_all(sock, length) if length > 0 else b""
    return msg_type, payload
